shared/health/src/main/java/io/pravega/shared/health/impl/CompositeHealthContributor.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.shared.health.impl;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.pravega.shared.health.ContributorRegistry;
+import io.pravega.shared.health.Health;
+import io.pravega.shared.health.HealthContributor;
+import io.pravega.shared.health.Status;
+import io.pravega.shared.health.StatusAggregator;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * The {@link CompositeHealthContributor} class defines the base logic required to build {@link HealthContributor} objects
+ * that depend on (or are composed of) other {@link HealthContributor} objects. To do this we need to due two things:
+ * reducing/reconciling the (potentially) many different {@link Status} states and a way to persist these relationships.
+ *
+ * {@link CompositeHealthContributor} objects are used *strictly* for grouping, meaning that they do not define their own
+ * health checking logic and therefore does not export any details ({@link Health#getDetails}).
+ */
+@Slf4j
+@ThreadSafe
+public abstract class CompositeHealthContributor implements HealthContributor {
+
+    /**
+     * The {@link StatusAggregator} used to perform the aggregation of all the {@link HealthContributor} dependencies.
+     */
+    @Getter
+    private final StatusAggregator aggregator;
+
+    private final ContributorRegistry registry;
+
+    private final Collection<HealthContributor> contributors = ImmutableSet.of();
```

Comment:
I don't get how this is useful. You are assigning an empty immutable list to this field - how is it used afterwards?

Reply:
The initial idea was to allow for the flexibility to have `HealthContributor` objects to be either registered under a `HealthRegistry` or within the `CompositeHealthContributor` itself. In retrospect this doesn't add much utility and unnecessarily increases complexity. Right now it serves the purpose for allowing a `CompositeHealthContributor` to be instantiated without having to be registered under a registry for the purposes of the 'health hierarchy' build/validation. See https://github.com/pravega/pravega/pull/5991/files#diff-6caef6c95eeb4891e53cf79c4d0811dcb8b4d2efe72cef05ca814711c7bd247bR180. The `HealthConfig` is being phased out to a different approach so this shouldn't be an issue moving forward.